### DISCUSSION TITLE
Adding emotions the tacotron2

### DISCUSSION THREAD

**Discussion by astricks** [coqui-ai/TTS#493] [Ideas]
I'm interested in adding emotion_id in addition to speaker_id so that while inference, I can choose which speaker as well as which emotion. The dataset would have to include an additional parameter (emotion_id) as well. This will enable me to do inference based on a specific speaker and specific emotion (angry, sad, happy, etc.)

I'm wondering how to go about doing this? Has something similar been done? Is this useful to anyone else?

**Comment by erogol** [comment]
You either need a dataset with emotion annotations and use these annotations as extra inputs to the model or if the dataset has no annotation but the speech is expressive then we can use sentiment classification to create these pseudo labels and train on them. 

Curious to hear that others say. So maybe there are more alternatives.

**Reply by astricks** [reply]
I indeed have an annotated TTS dataset - single speaker, multiple emotions. As an experiment, I'm trying to just use the multi-speaker embedding and pass each emotion has a different speaker. I'll report back with updates.

**Reply by erogol** [reply]
ohh it's a good idea. So basically you use the multi-speaker setting for multi-emotions trainining :).

**Reply by a-froghyar** [reply]
@astricks sounds like a cool idea, let us know if it works out!

**Reply by thorstenMueller** [reply]
Just in case you want to play around with another emotional dataset ;-).
https://github.com/thorstenMueller/deep-learning-german-tts/#dataset-Thorsten-emotional

**Reply by astricks** [reply]
Ok, good news, it seems to be working. Also since it's the same speaker speaking different emotions, the model aligns decently and I'm able to tell the difference in emotion. I'll train it for another 300k/400k steps and share some voice samples.

**Reply by astricks** [reply]
It certainly is having a harder time finding alignment as compared to the single speaker model where I combine all the emotions. After 800k steps, the model isn't performing as well as the single speaker model. I have 21 hours of data from the single speaker, but when I divide it by emotion, it is only 3 hours per emotion (speaker). I was hoping that wouldn't be an issue since all 7 emotions are technically the same speaker. I'm recording more data to see if that helps. I also warm-started using the single-speaker model I had trained earlier. It seems to be improving marginally, maybe it gets better after 1M steps?

**Question** : I'm wondering if warm-starting off the single speaker model to fine-tune 7 single-speaker models, one per emotion. Would that help? The only difference would be not using speaker_embedding, I'd imagine.

**Reply by thorstenMueller** [reply]
Hi @astricks .
I'm afraid i can't answer your question, but due you experimented with my single speaker emotional dataset i want to say that i released version 2 yesterday. I added "drunk" and "whispering" recordings.

**Reply by Sadam1195** [reply]
This might be not specifically related but could be answer to your question. You can improve things until they reach to point of diminishing return after that no matter how much you train the model mostly likely the chances are you won't get things improve more significantly. Your bottle neck could be the data itself (quality or quantity) or the model you have choose to train. @astricks

**Reply by astricks** [reply]
I'm going to gather more data and re-try. Will share back results in a month.

**Reply by astricks** [reply]
Update - Instead of waiting for more data, I gave it another sincere try. This time, it worked.

This time,
1. I used the speaker encoder to calculate speaker embeddings and trained using the resulting external_speaker_encoder_file
2. I diligently reduced r and the learning rate

Here are some samples
[Angry](https://drive.google.com/file/d/1IKECSr7RNnhseTggPCJ0AflfM73xzAio/view?usp=sharing) 
[Scared](https://drive.google.com/file/d/1bUPP5aLloDIDpMFQa_bfRcRY50mu9Z-R/view?usp=sharing) 
[Surprised](https://drive.google.com/file/d/155hWdchDBZNksBfl4TalhhnZLp_fFZ1q/view?usp=sharing) 

This was an interesting exercise. Thank you all for your helpful comments.

**Reply by erogol** [reply]
@astricks good results and interesting use of the speaker encoder.

Maybe @thorstenMueller  can try the same approach. 💪

**Reply by ridhisoni7** [reply]
@astricks do you have the code from this experiment up anywhere? would be very helpful for a project im working on atm! thanks :)
